internal/core/swrenderer/draw_functions.rs

```diff
@@ -231,81 +235,105 @@ pub(super) fn draw_rounded_rectangle_line(
-        TargetPixel::blend_pixel(&mut line_buffer[pos_x + x], Premultiplied::premultiply(col))
+        let col = PremultipliedRgbaColor {
+            alpha: (((c.alpha as u32) * (255 - cov) as u32) / 255) as u8,
+            red: (((c.red as u32) * (255 - cov) as u32) / 255) as u8,
+            green: (((c.green as u32) * (255 - cov) as u32) / 255) as u8,
+            blue: (((c.blue as u32) * (255 - cov) as u32) / 255) as u8,
+        };
+        line_buffer[pos_x + x].blend(col);
     });
 }
 
-// a is between 0 and 255. When 0, we get color1, when 2 we get color2
-fn interpolate_color(a: u32, color1: Color, color2: Color) -> Premultiplied {
+// a is between 0 and 255. When 0, we get color1, when 255 we get color2
+fn interpolate_color(
+    a: u32,
+    color1: PremultipliedRgbaColor,
+    color2: PremultipliedRgbaColor,
+) -> PremultipliedRgbaColor {
     let b = 255 - a;
 
-    let al1 = color1.alpha() as u32;
-    let al2 = color2.alpha() as u32;
+    let al1 = color1.alpha as u32;
+    let al2 = color2.alpha as u32;
 
     let a_ = a * al2;
     let b_ = b * al1;
     let m = a_ + b_;
 
     if m == 0 {
-        return Premultiplied(Color::default());
+        return PremultipliedRgbaColor::default();
     }
 
-    let div = 255 * 255;
-    Premultiplied(Color::from_argb_u8(
-        (m / 255) as u8,
-        ((b_ * color1.red() as u32 + a_ * color2.red() as u32) / div) as u8,
-        ((b_ * color1.green() as u32 + a_ * color2.green() as u32) / div) as u8,
-        ((b_ * color1.blue() as u32 + a_ * color2.blue() as u32) / div) as u8,
-    ))
+    PremultipliedRgbaColor {
+        alpha: (m / 255) as u8,
+        red: ((b * color1.red as u32 + a * color2.red as u32) / 255) as u8,
+        green: ((b * color1.green as u32 + a * color2.green as u32) / 255) as u8,
+        blue: ((b * color1.blue as u32 + a * color2.blue as u32) / 255) as u8,
+    }
+}
+
+/// A color whose component have been pre-multiplied by alpha
+///
+/// The renderer operates faster on pre-multiplied color since it
+/// caches the multiplication of its component
+///
+/// PremultipliedRgbaColor can be constructed from a [`Color`] with
+/// the [`From`] trait. This conversion will pre-multiply the color
+/// components
+#[derive(Clone, Copy, Debug, Default)]
+pub struct PremultipliedRgbaColor {
+    pub alpha: u8,
+    pub red: u8,
+    pub green: u8,
+    pub blue: u8,
 }
 
-/// Wrap a color whose component have been pre-multiplied by alpha
-#[derive(Clone, Copy)]
-pub struct Premultiplied(pub Color);
+/// Convert a non-premultiplied color to a premultiplied one
+impl From<Color> for PremultipliedRgbaColor {
+    fn from(col: Color) -> Self {
+        Self::premultiply(col)
+    }
+}
 
-impl Premultiplied {
+impl PremultipliedRgbaColor {
     /// Convert a non premultiplied color to a premultiplied one
-    pub fn premultiply(col: Color) -> Self {
+    fn premultiply(col: Color) -> Self {
```

Comment:
I wonder if it would perhaps be more idiomatic to call this function “new”?

Reply:
I like to be explicit here that this do the pre-multiplication
This is a private function anyway. The public interface is through the From trait

Reply:
Fair point :)